Add state machine page - Closes #1495 (#1504)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/understand-blockchain/state-machine.adoc

= State machine
//Settings
:toc:
:fn_lip32: footnote:lip32[the details are specified in https://github.com/LiskHQ/lips/blob/main/proposals/lip-0032.md[LIP 0032^].]

// URLs
:url_wiki_dfsm: https://en.wikipedia.org/wiki/Deterministic_finite_automaton
:url_blog_tree: https://lisk.com/blog/research/introducing-lisk-tree
:url_blog_merkle: https://lisk.com/blog/research/sparse-merkle-trees-and-new-state-model

Every blockchain is at its core a replicated {url_wiki_dfsm}[deterministic finite-state machine^].

> * A *state machine* is a concept in computer science, consisting of the following elements:

Reviewer: Not sure what the * should correspond to, but the rendered version shows nothing (no bold nor italics)

Reviewer: ok

Reviewer: actually it was blocked by 'noscript' 😅 all good

> ** *states:* A state machine typically has a set of different states, that it can enter.

Reviewer: ```suggestion
** *states:* A state machine typically has a set of different states that it can enter.
```

For example, a padlock represented as state machine would have the two states "Open" and "Locked".
** *transitions:* a set of allowable operations that change from one state to another.
For example, to change from state "Open" to "Locked", a transition would be "Close padlock", and from "Locked" to "Open", it would be "Insert key & turn".

> * *Replicated* describes the characteristic of blockchains to consist of a network of nodes all storing an individual copy of the blockchain, which replicates the state machine on each connected node.

Reviewer: Here I would describe what 'replicated' means in this more general context (i.e., the state machine is replicated across many nodes. Each node maintains independently the current state of the state machine). We could also mention that this only works for deterministic state machine

> * *Finite* means, the number of states, that the machine can reach is always finite.

Reviewer: ```suggestion
* *Finite* means that the number of states that the machine can reach is finite (not infinite).
```

The dataset cannot be altered without changing the Merkle root.
* *tree:* The underlying data structure is organized as a tree, where each parent node is obtained by hashing the data from the child nodes in the layer below.
Here we consider only *binary trees*, where each parent has two children.

A *sparse Merkle tree (SMT)* is very similar to a regular Merkle Tree, with the following difference:
The order of insertion of data blocks does not matter, and the root of the tree only depends on the final state of the map.
That means, in an SMT, the input of the hash function is the underlying <<kv-maps, key-value map>> and the output is the Merkle root.
Due to the history independence, SMTs are the suitable choice to authenticate key-value maps.

.RMT vs SMT
image::understand-blockchain/rmt-vs-smt.png[]

A blockchain created with the Lisk SDK defines a single global store.
An SMT is created on top of the global state store, called the **state tree**.
Then the Merkle root of this tree, the state root, is inserted into each block header.

Each module registered in the chain defines its own generic key-value map, where:

* *key* = Module ID | Store prefix | Store key

> ** **Module ID** = The ID of the module, identifying the key-value map. We will reserve all module IDs starting with a 0 in their binary representation to modules that are shipped with the SDK by default. The IDs of modules developed as part of each blockchain application will start with a 1. The module ID has a fixed length of 4 bytes.

Reviewer: ```suggestion
** **Module ID**: The ID of the module, identifying the key-value map. We will reserve all module IDs starting with a 0 in their binary representation to modules that are shipped with the SDK by default. The IDs of modules developed as part of each blockchain application will start with a 1. The module ID has a fixed length of 4 bytes.
```

Reviewer: I would use colons for consistency with the previous section

image::understand-blockchain/smt.png[]

== Use cases of SMTs

A list of all use cases for SMTs in the Lisk Protocol, and their corresponding SMT.

.Use cases for SMTs in the Lisk protocol
[cols="2,2a,5a"]
|===
|Use Case|SMT|Description

|Interoperability|`stateRoot`|The `stateRoot` is used during the processing of cross-chain update transactions to authenticate the outgoing messages from one chain.
|State consistency|`stateRoot`|All nodes in the network will be able to check the consistency of the state just by checking the `stateRoot`.
For instance, if a block contains a transaction that randomly credits either a user A or a user B with tokens, then any node obtaining a different result from the block forger would immediately reject the block, as the two state roots would differ.
|Light clients|`stateRoot`|The `stateRoot` can be used to provide efficient proofs for the state of a certain account.
|Decentralized Regenesis|`stateRoot`|A snapshot of the blockchain can be used to perform a hardfork to implement the new protocol.
Reference to the last blockchain state will be stored as the Merkle root of the hashes of all blocks up to the snapshot.

> |Proofs-of-Inclusion for Transactions in a Block|`transactionRoot`|Each block header stores the `transactionRoot`.

Reviewer: We use regular Merkle trees for the transaction root

The `transactionRoot` is calculated as the Merkle root of the IDs of the transactions included in the block payload{fn_lip32}.
Using the `transactionRoot` and a proof-of-inclusion, it is be possible to check whether a certain transaction is part of the block without downloading the full block.

> |Proofs-of-Inclusion for Assets in a Block|`assetRoot`| The root of the event sparse Merkle tree is added to the block header to allow checking inclusion and non-inclusion of block assets.

Reviewer: Copy/paste error? the assetRoot is a regular Merkle tree root, I would just remove this line

Author: I actually was not sure, if we still use RMTs in the Lisk Protocol. In that case, I will mention them seperately, I think it is an interesting information.

In Lisk, a *key-value-store* is used as database, containing all data that is considered part of the blockchain.
* A *transition* of states is triggered through *blocks* and their containing *transactions* which are added to the blockchain.
When executing a block `B`, the state-machine changes its state from a state `S` to a state `S’`.
This one state transition can further be broken down into smaller intermediate state transitions, for instance, by considering that executing a block means executing all transactions in that block sequentially.

== The blockchain state as Sparse Merkle Tree

A Merkle tree is an authenticated data structure organized as a tree.
Let's try to understand what exactly each of these words mean:

* *Merkle:* This name is derived from the actual inventor of hash trees, Ralph Merkle.
* *data structure:* A data structure is a collection of data with a set of operations that can be performed on it.
An example of a data structure is an array, an object, or a Merkle tree.
* *authenticated:* The integrity of the data structure can be efficiently verified using the (Merkle) root of the tree.
The dataset cannot be altered without changing the Merkle root.

> * *tree:* The underlying data structure is organized as a tree, where each parent node is obtained by hashing the data from the child nodes in the layer below.

Reviewer 2: maybe these can also start with capital letter

|Light clients|`stateRoot`|The `stateRoot` can be used to provide efficient proofs for the state of a certain account.
|Decentralized Regenesis|`stateRoot`|A snapshot of the blockchain can be used to perform a hardfork to implement the new protocol.
Reference to the last blockchain state will be stored as the Merkle root of the hashes of all blocks up to the snapshot.
|Proofs-of-Inclusion for Events in a Block|`eventRoot`| The root of the event sparse Merkle tree is added to the block header to allow checking inclusion and non-inclusion of events.
|===

A list of all use cases for RMTs in the Lisk Protocol, and their corresponding RMT.

.Use cases for RMTs in the Lisk protocol
[cols="2,2a,5a"]
|===
|Use Case|RMT|Description
|Proofs-of-Inclusion for Transactions in a Block|`transactionRoot`|Each block header stores the `transactionRoot`.
The `transactionRoot` is calculated as the Merkle root of the IDs of the transactions included in the block payload{fn_lip32}.
Using the `transactionRoot` and a proof-of-inclusion, it is be possible to check whether a certain transaction is part of the block without downloading the full block.

> |Proofs-of-Inclusion for Assets in a Block|`assetRoot`| The root of the asset Merkle tree is added to the block header to allow checking inclusion and non-inclusion of block assets.

Reviewer 2: Maybe Proofs-of-(non-)Inclusion for Assets in a Block as check both of them
